soroban-sdk/src/contract_event.rs

```diff
@@ -24,43 +32,91 @@ use crate::{env::internal, Env, IntoVal, RawVal, Vec};
 /// # #[cfg(feature = "testutils")]
 /// # fn main() {
 /// #     let env = Env::default();
-/// #     let contract_id = FixedBinary::from_array(&env, [0; 32]);
+/// #     let contract_id = BytesN::from_array(&env, [0; 32]);
 /// #     env.register_contract(&contract_id, Contract);
 /// #     f::invoke(&env, &contract_id);
 /// # }
 /// # #[cfg(not(feature = "testutils"))]
 /// # fn main() { }
 /// ```
 #[derive(Clone)]
-pub struct ContractEvent(Env);
+pub struct Events(Env);
 
-impl Debug for ContractEvent {
+impl Debug for Events {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "ContractEvent")
+        write!(f, "Events")
+    }
+}
+
+pub trait IntoTopics {
+    fn into_topics(self, env: &Env) -> Object;
```

Comment:
I think IntoTopics could probably just be a marker trait, it doesn't actually need a function or any functionality.


i.e.
```suggestion
pub trait Topics {
```

Then the macro impl for the tuples is just doing:

```rust
impl<T0, T1, ...> Topics for (T0, T1, ...) {}
```

And the publish function can be defined as:

```rust
     pub fn publish<T, D>(&self, topics: T, data: D)
     where
         T: Topics + IntoVal<Env, RawVal>,
```

Reply:
I like this suggestion! I've changed the `IntoTopics` to `Topics: IntoVal<Env, Object>` as you suggested. The `IntoVal` part replaces the need for the previous `into_topics` function.
I had to adapt the `Vec<T>` and `()` to impl the `IntoVal` trait. The `Vec` I think is fine. But `()` I'm a little unsure this is right place where it should go.

Reply:
👀